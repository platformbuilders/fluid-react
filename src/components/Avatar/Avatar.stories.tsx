import { expect, jest } from '@storybook/jest';
import type { Meta, StoryObj } from '@storybook/react';
import { configure, userEvent, within } from '@storybook/testing-library';
import Avatar from './index';

configure({
  testIdAttribute: 'id',
});

//  Mocks
const eventMockFn = jest.fn();
const testId = 'avatar-id';

const meta: Meta<typeof Avatar> = {
  title: 'Components/Avatar',
  component: Avatar,

  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    src: 'https://thispersondoesnotexist.com/',
    onPress: eventMockFn,
    id: testId,
    style: { width: '150px', height: '150px', objectFit: 'cover' },
  },

  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);

    await step('Avatar - Normal | Test Render', async () => {
      expect(canvas.getByTestId(testId)).toBeInTheDocument();
    });

    await step('Avatar - Normal | Check Event Click', async () => {
      await userEvent.click(canvas.getByTestId(testId));

      expect(eventMockFn).toHaveBeenCalled();
    });
  },
};

type Story = StoryObj<typeof Avatar>;

export const Rounded: Story = {
  args: {
    variant: 'rounded',
  },
};

export const Circle: Story = {
  args: {
    variant: 'circle',
  },
};

export const Square: Story = {
  args: {
    variant: 'square',
  },
};

export default meta;
