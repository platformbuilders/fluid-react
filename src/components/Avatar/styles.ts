import styled from 'styled-components';
import { RNCamera } from 'react-native-camera';
import FastImage from 'react-native-fast-image';
import Touchable from '../Touchable';
import { ifStyle } from '../../utils/helpers';

const showBorder = ifStyle('showBorder');

export const Wrapper = styled(Touchable)<{
  size: number;
  disabled: boolean;
  showBorder: boolean;
}>`
  width: ${(props): number => props.size}px;
  height: ${(props): number => props.size}px;
  border-radius: ${(props): number => props.size / 2}px;
  overflow: hidden;
  border: ${showBorder('4px solid white', '')};
`;

export const Image = styled(FastImage).attrs({ resizeMode: 'cover' })<{
  source: any;
}>`
  width: 100%;
  height: 100%;
`;

export const CameraView = styled(RNCamera)<{ size: number }>`
  width: ${(props): number => props.size}px;
  height: 100%;
  border-radius: ${(props): number => props.size / 2}px;
  overflow: hidden;
  border: ${showBorder('4px solid white', '')};
`;
