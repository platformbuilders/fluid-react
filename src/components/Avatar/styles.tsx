import styled from 'styled-components';
import MaterialAvatar from '@material-ui/core/Avatar';
import Touchable from '../Touchable';

export const DefaultAvatar = styled(({ onPress, accessibility, ...rest }) => (
  <Touchable
    accessibility={accessibility || 'Foto ou Avatar'}
    onPress={onPress}
  >
    <MaterialAvatar {...rest} />
  </Touchable>
))``;
