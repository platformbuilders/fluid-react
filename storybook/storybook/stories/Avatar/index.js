import * as React from 'react';
import { Avatar as BuildersAvatar } from 'builders-ui';
import CenteredView from '../utils/CenteredView';

const Avatar = () => {
  return (
    <CenteredView>
      <BuildersAvatar
        image={
          'https://media.licdn.com/dms/image/C560BAQED6xsxb4hwOA/company-logo_200_200/0?e=2159024400&v=beta&t=xJU92Gb_eWF3dalwZEaubtm7ILRa0O101xIO2diPkZQ'
        }
        showBorder
        size={100}
        onPress={() => {}}
        onUpload={() => {}}
      />
    </CenteredView>
  );
};

export default Avatar;
