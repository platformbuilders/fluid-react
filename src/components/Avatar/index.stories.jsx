import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import markdown from './avatar.md';
import Avatar from './index';

storiesOf('Avatar', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      content: markdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => (
    <Avatar
      onPress={action('clicked')}
      src={text(
        'src',
        'https://api.adorable.io/avatars/285/abott@adorable.png'
      )}
    />
  ));
