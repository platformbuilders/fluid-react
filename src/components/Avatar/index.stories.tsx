import type { Meta, StoryObj } from '@storybook/react';
import Avatar from './index';

const meta: Meta<typeof Avatar> = {
  title: 'Avatar',
  component: Avatar,
  args: {
    onPress: () => console.log('press'),
    src: 'https://toppng.com/uploads/preview/avatar-png-11554021661asazhxmdnu.png',
  },
};

export default meta;
type Story = StoryObj<typeof Avatar>;

export const Default: Story = {};
