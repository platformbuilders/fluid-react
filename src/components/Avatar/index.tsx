import { FC } from 'react';
import { DefaultAvatar } from './styles';

export type AvatarFormats = 'circle' | 'rounded' | 'square';

export type Props = {
  className?: string;
  src: string;
  alt?: string;
  variant?: AvatarFormats;
  onPress(): void;
};

const Avatar: FC<Props> = ({
  className,
  src,
  alt = '',
  variant = 'circle',
  onPress,
  ...rest
}): JSX.Element => (
  <DefaultAvatar
    className={className}
    alt={alt}
    src={src}
    variant={variant}
    onPress={onPress}
    {...rest}
  />
);

export default Avatar;
