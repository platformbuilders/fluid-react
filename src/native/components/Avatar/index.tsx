import React, { useState, useRef, useImperativeHandle } from 'react';
import { isEmpty } from 'lodash';
import { RNCamera } from 'react-native-camera';
import FastImage, { Source } from 'react-native-fast-image';
import ImagePicker from 'react-native-image-picker';
import { Wrapper, CameraView, UploadIconWrapper, UploadIcon } from './styles';

const defaultAvatar =
  'https://static1.joj.sk/html/assets/avatar-placeholder.jpg';

type Props = {
  ref?: any;
  image?: string;
  showBorder?: boolean;
  borderWidth?: number;
  borderColor?: string;
  displayCamera?: boolean;
  size?: number;
  onPress?: (x: any) => void;
  onUpload?: (x: any) => any;
  id: string;
  accessibility: string;
  accessibilityLabel?: string;
  testID?: string;
};

const Avatar: React.FC<Props> = React.forwardRef(
  (
    {
      id,
      image = defaultAvatar,
      accessibility = 'Upload de Avatar',
      accessibilityLabel,
      testID,
      size = 50,
      onPress,
      onUpload,
      showBorder = true,
      displayCamera = false,
      ...rest
    },
    ref,
  ) => {
    const [uploadedImage, setUploadedImage] = useState<any>();
    const cameraRef = useRef<any>();

    const openPicker = (): Promise<void> => {
      const options = {
        title: 'Selecionar foto',
        cancelButtonTitle: 'Cancelar',
        takePhotoButtonTitle: 'Tirar foto',
        chooseFromLibraryButtonTitle: 'Escolher da galeria',
        quality: 0.3,
        storageOptions: {
          skipBackup: true,
          path: 'images',
        },
      };
      return new Promise((resolve) => {
        ImagePicker.launchImageLibrary(options, (response) => {
          setUploadedImage(response.uri);
          if (onUpload) {
            onUpload(response.uri);
          }
          resolve();
        });
      });
    };

    const takePicture = async (): Promise<void> => {
      if (cameraRef.current) {
        const options = { quality: 0.5, base64: true };
        const data = await cameraRef.current.takePictureAsync(options);
        setUploadedImage(data);
        return data;
      }
    };

    const getUploadImage = (): any => {
      return uploadedImage;
    };

    const clearUploadImage = (): void => {
      setUploadedImage('');
    };

    const getCurrentAvatar = (): Source | any => {
      if (uploadedImage) {
        return { uri: uploadedImage };
      }
      if (image && !isEmpty(image)) {
        return { uri: image };
      }
      return defaultAvatar;
    };

    useImperativeHandle(ref, () => ({
      getUploadImage,
      clearUploadImage,
      takePicture,
      openPicker,
    }));

    return (
      <Wrapper
        id={id}
        accessibility={accessibility}
        accessibilityLabel={accessibilityLabel || accessibility}
        testID={testID || id}
        size={size}
        onPress={onPress}
        disabled={!onPress}
        showBorder={showBorder}
        {...rest}
      >
        {displayCamera && !uploadedImage ? (
          <CameraView
            ref={cameraRef}
            size={size}
            type={RNCamera.Constants.Type.front}
            flashMode={RNCamera.Constants.FlashMode.auto}
            androidCameraPermissionOptions={{
              title: 'Câmera',
              message: 'Precisamos da sua permissão para usar a câmera',
              buttonPositive: 'Ok',
              buttonNegative: 'Cancelar',
            }}
          />
        ) : (
          <FastImage
            source={getCurrentAvatar()}
            resizeMode={FastImage.resizeMode.cover}
            style={{ width: '100%', height: '100%' }}
          />
        )}
        <UploadIconWrapper size={size}>
          <UploadIcon id="" accessibility="" />
        </UploadIconWrapper>
      </Wrapper>
    );
  },
);

Avatar.displayName = 'AvatarComponent';

export default Avatar;
