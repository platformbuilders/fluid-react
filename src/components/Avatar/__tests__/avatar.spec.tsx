import React from 'react';
import faker from 'faker';
import { shallow } from 'enzyme';

import Avatar, { Props } from '..';

let defaultProps: Props = {
  src: 'some_url',
  onPress: jest.fn(),
};

describe('Component: Avatar', () => {
  beforeEach(() => {
    defaultProps = { ...defaultProps };
  });

  test('snapshots with default props', () => {
    const component = shallow(<Avatar {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  test('snapshots with other props', () => {
    const component = shallow(
      <Avatar {...defaultProps} alt="alt text" variant="square" />,
    );

    expect(component).toMatchSnapshot();
  });

  test('check props', () => {
    // should
    const altMock = faker.random.words();
    const srcMock = faker.random.words();
    const variantMock = 'rounded';

    // when
    const component = shallow(
      <Avatar
        onPress={jest.fn}
        src={srcMock}
        alt={altMock}
        variant={variantMock}
      />,
    );

    // then
    expect(component.props().src).toEqual(srcMock);
    expect(component.props().alt).toEqual(altMock);
    expect(component.props().variant).toEqual(variantMock);
  });

  test('should call onPress when pressed', () => {
    // shuold
    const onPressMock = jest.fn();
    const component = shallow(
      <Avatar {...defaultProps} onPress={onPressMock} />,
    );

    // when
    component.simulate('press');

    // then
    expect(onPressMock).toHaveBeenCalled();
  });
});
