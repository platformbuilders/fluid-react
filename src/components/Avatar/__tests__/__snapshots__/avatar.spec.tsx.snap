// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Avatar snapshots with default props 1`] = `
<div>
  <span>
    <span />
  </span>
</div>
`;

exports[`Component: Avatar snapshots with other props 1`] = `
<div>
  <span>
    <span />
  </span>
</div>
`;
